Implement ERR_raise() / ERR_raise_data()

levitte opened the pull request:
`ERR_raise()` and `ERR_raise_data` is new functionality to raise errors, with this intended signature:

``` C
void ERR_raise(int lib, int reason);
void ERR_raise_data(int lib, int reason, const char *fmt, ...);
```

The intention with this is to have functionality that makes it easier to report errors with additional text, instead of the sometimes cumbersome error handling with `ERR_set_error_data()`.

`ERR_raise()` and `ERR_raise_data` are implemented as macros which works with a few building blocks:

- `ERR_new()` which simply allocates a new error record in the error queue.
- `ERR_set_debug()` which sets debugging information in the current error record.  "debugging information" are `__FILE__`, `__LINE__` and `__func__` or variants thereof (in reality, we use `OPENSSL_FILE`, `OPENSSL_LINE` and `OPENSSL_FUNC`, which are defined appropriately for the toolchain that builds this)
- `ERR_set_error()` which sets the basic error information in the current error record; library number, reason code, and a format string followed by arguments that form the additional data.

All the `FOOerr` macros are reimplemented in terms of `ERR_raise()` / `ERR_raise_data`, including `FUNCerr`, which also means that `ERR_put_func_error()` lost its meaning.

-----

This functionality was briefly discussed in [this thread on openssl-project](https://mta.openssl.org/pipermail/openssl-project/2019-June/001426.html)

levitte committed:
Move some macros from include/openssl/opensslconf.h.in, add OPENSSL_FUNC

New header file, include/openssl/macros.h, which contains diverse
useful macros that we use elsewhere.

We also add the new macro OPENSSL_FUNC, which is an alias for
__FUNC__, __FUNCTION__, __FUNCSIG or __func__, depending on what the
compiler supports.  In the worst case, it's an alias for the string
"(unknown function)".

levitte committed:
ERR: refactor useful inner macros to err_locl.h.  Add function name field

The useful inner macros are now static inline functions.  That will
make them easier to debug in the future.

levitte committed:
ERR: Add new building blocks for reporting errors

The new building block are ERR_new(), ERR_set_debug(),
ERR_set_error(), ERR_vset_error(), which allocate a new error record
and set the diverse data in them.  They are designed in such a way
that it's reasonably easy to create macros that use all of them but
then rely completely on the function signature of ERR_set_error() or
ERR_vset_error().

levitte committed:
ERR: Implement the macro ERR_raise() and use it

This macro uses a trick in C.  The following is permitted:

    #include <stdio.h>

    void first(void)
    {
        printf("Hello! ");
    }

    void foo(const char *bar)
    {
        printf("%s", bar);
    }

    int main()
    {
        /* This */
        (first(),foo)("cookie");
    }

ERR_raise() can be used to implement FUNCerr() as well, which takes
away the need for the special function ERR_put_func_error().

squash! ERR: Implement the macro ERR_raise() and use it

Make sure to mention ERR_raise_data() too

levitte committed:
ERR: Remove ERR_put_func_error() and reimplement ERR_put_error() as a macro

Also, deprecate ERR_put_error()

levitte committed:
Refactor provider support for reporting errors

The core now supplies its own versions of ERR_new(), ERR_set_debug()
and ERR_vset_error().  This should suffice for a provider to have any
OpenSSL compatible functionlity it desires.

The main difference between the ERR functions and the core
counterparts is that the core counterparts take an OSSL_PROVIDER
parameter instead of the library number.  That way, providers do not
need to know what number they have been assigned, that information
stays in the core.

levitte committed:
Adapt the FIPS provider to use the new core error functions

levitte commented:
One commit mentions the following little program

``` C
#include <stdio.h>

void first(void)
{
    printf("Hello! ");
}

void foo(const char *bar)
{
    printf("%s", bar);
}

int main()
{
    /* This */
    (first(),foo)("cookie");
}
```

It shows the trick used to implement `ERR_raise()`.  I have tested that little program on Linux (using gcc), Windows (using MSVC) and VMS (using DEC C)

t8m commented:
Dropping ERR_put_error() is an API break.

t8m commented:
> Dropping ERR_put_error() is an API break.

Ah, I did not notice that it is now implemented as macro. OK then.

richsalz commented on include/openssl/err.h:
```diff
@@ -106,49 +107,49 @@ typedef struct err_state_st {
-# define PROVerr(f,r) ERR_PUT_error(ERR_LIB_PROV,0,(r),OPENSSL_FILE,OPENSSL_LINE)
+#  define SYSerr(f,r)  ERR_raise(ERR_LIB_SYS,(r),NULL)
+# endif
+# define FUNCerr(f,r)  ERR_raise(ERR_LIB_SYS,(r),"calling function %s",(f))
```
should these just be alpha-sorted and not, time-of-addition :) ordered?

levitte replied:
The are sorted in library number order, are they not?

richsalz replied:
It is more useful to humans to be in alphabetical order.

richsalz reviewed (commented):
So this makes #9441 needless? Or do folks still want the unused argument removed?

I am concerned that this could add variable timing to errors and therefore be a timing oracle. I'd be delighted to be convinced that's needless, can you do that?

levitte commented:
> So this makes #9441 needless? Or do folks still want the unused argument removed?

Depends...  do people still want to use the `FOOerr` / `FOOerror` type of macro, or will they turn to `ERR_raise`?  I don't see this competing directly with #9441, but I agree that if we make `ERR_raise` the norm, then we have no more use of `FOOerr` (and concequently, `FOOerror`), they will merely become a historical note, and I understand how `FOOerror` would be pointless in that case.

> I am concerned that this could add variable timing to errors and therefore be a timing oracle.

Only if you have additional data, *which is a problem we already have with `ERR_add_error_data`*

levitte commented:
Regarding the timing oracle and additional error data...  we do have a problem here, 'cause our error message are quite terse and non-descript at times, so there's a desire to add additional data, but the clumsiness of `ERR_add_error_data` has perhaps made that less encouraged.

richsalz commented:
If this gets merged, then I will look at changing the script in #9441 to rewrite things in terms of ```ERR_raise```  Or I will close the PR because adding 43 macros that are only temporary "glue" is a bad idea.

levitte committed:
Avoid using ERR_put_error() directly in OpenSSL code

If compiled with 'no-deprecated', ERR_put_error() is undefined.  We
had one spot where we were using it directly, because the file and
line information was passed from elsewhere.

Fortunately, it's possible to use ERR_raise() for that situation, and
call ERR_set_debug() immediately after and thereby override the
information that ERR_raise() stored in the error record.

squash! Avoid using ERR_put_error() directly in OpenSSL code

util/mkerr.pl needed a small adjustment to not generate code that
won't compile in a 'no-deprecated' configuration.

paulidale commented on include/openssl/err.h:
```diff
@@ -234,9 +235,28 @@ typedef struct ERR_string_data_st {
+void ERR_vset_error(int lib, int reason, const char *fmt, va_list args);
+
+/* Main error raising functions */
+#define ERR_raise                                               \
```
Neat construction.

I don't use the `,` comma operator enough...

paulidale commented:
This looks like a good approach.

richsalz commented on crypto/err/err_blocks.c:
```diff
@@ -0,0 +1,74 @@
+        return;
+
+    if (fmt != NULL
+        && (buf = OPENSSL_malloc(ERR_MAX_DATA_SIZE)) != NULL) {
```
Yes, this is very bad.

levitte replied:
Note that this is additional data, i.e the same kind of stuff you get with `ERR_add_error_data`. But yes, that can be done.
The reason string works as before

levitte replied:
Actually, I changed my mind, I see no reason to add anything special unless you're telling me that `ERR_add_error_vdata()` should as well.

richsalz commented on crypto/err/err_blocks.c:
```diff
@@ -0,0 +1,74 @@
+        buf[buf_len] = '\0';
+
+        /* Try to reduce the size */
+        buf = OPENSSL_realloc(buf, buf_len + 1);
```
calling vnprintf twice seems much better solution.

levitte replied:
Does `BIO_vsnprintf` even have that sort of functionality, that one can do the "i2d trick"?

levitte replied:
Having looked further, I'm not sure what you're suggesting...  are you suggesting I start with a small buffer, and then what, that if that small buffer is filled entirely, I try again with a larger buffer?  That sounds enormously complex in comparison.

However, I could go for a smaller buffer.  I would agree that 4K is quite beefy (that's 50+ lines on a 80-column terminal) and could go down to 1K (12 lines of additional data is already quite a lot)

richsalz commented on crypto/err/err_locl.h:
```diff
@@ -0,0 +1,52 @@
+ * https://www.openssl.org/source/license.html
+ */
+
+#define err_allocate(p)                                         \
```
these are not time-critical, just make them internal functions.  Then you don't have to worry about parenthesizing or someone calling ```err_clear_data(p, i++)```

richsalz replied:
```err_allocate``` is a misleading name since it doesn't actually allocate anything.

levitte replied:
But it does allocate a new error record slot. Sure, it's in a ring buffer, but still. Allocation as a term covers more than merely allocating memory on the heap, doesn't it? I'm pretty sure I have seen that word used more broadly.

However, feel free to suggest a name that you find more suitable.

richsalz replied:
```err_get_slot``` ?

levitte replied:
Sure

paulidale replied:
Is the name changing for this one?

I tend to agree that (internal) functions might be a nicer choice.

richsalz commented on crypto/property/property_parse.c:
```diff
@@ -358,14 +347,13 @@ OSSL_PROPERTY_LIST *ossl_parse_property(OPENSSL_CTX *ctx, const char *defn)
-            PROPerr(PROP_F_OSSL_PARSE_PROPERTY, PROP_R_PARSE_FAILED);
-            ERR_add_error_data(2, "Unknown name HERE-->", start);
+            ERR_raise(ERR_LIB_PROP, PROP_R_PARSE_FAILED,
+                      "Unknown name HERE-->%s", start);
```
```"Unknown name \"%s\""``` is much better than ```HERE-->```
In all places where that construct is used.

levitte replied:
I have kept the existing message. Also, the pointer doesn't show just the name, but everything that follows in the property string. I can't insert a NUL byte in there

richsalz commented on crypto/err/err_blocks.c:
```diff
@@ -0,0 +1,74 @@
+        return;
+
+    if (fmt != NULL
+        && (buf = OPENSSL_malloc(ERR_MAX_DATA_SIZE)) != NULL) {
```
This approach, alloc potential worst-case then shrink it, has two failure points. If malloc fails use fixed-string like "extra data suppressed, allocation failure" or some such. If you are that concerned about the size, call vsnprintf twice, or start at 1K and go up.

levitte replied:
I've reduced the size to 1K

richsalz replied:
I still think that is way to big and suggest 256 (sorry 1K looked big compared to 4K); four lines of typical text are still way more than we need to support.  Don't realloc, let vsnprintf do its truncation work.  And keep/re-use the pointer.

richsalz commented on crypto/err/err_blocks.c:
```diff
@@ -0,0 +1,74 @@
+        && (buf = OPENSSL_malloc(ERR_MAX_DATA_SIZE)) != NULL) {
+        int printed_len = 0;
+
+        printed_len = BIO_vsnprintf(buf, ERR_MAX_DATA_SIZE, fmt, args);
```
If there's already a buffer, you should try to re-use it.  And that means keeping the buffer and size in the error-state.

levitte replied:
I'll look into that.

levitte replied:
I did a quick look, and I can only conclude that I have no idea what you're talking about.  "keeping the buffer and size in the error state" doesn't mean anything, as *this processes additional data* (i.e. the same stuff that  `ERR_add_error_data` deals with)

richsalz replied:
The ```err_data``` field might already point to memory, and that memory might be big enough.  Does that help?

levitte replied:
How do I know that it's big enough?

levitte replied:
The current modus operandi is that when a new slot gets allocated, all its data is cleared (i.e. the `err_data` field is freed).

It's quite possible that we could make that more gentle on allocation/deallocation, but I'm starting to feel that it's beyond what I'm willing to put into *this* PR.  You have me thinking, though...  maybe I'll do something about that, in a separate PR...  or well, PR welcome, if anyone feels inclined!

richsalz replied:
> How do I know that it's big enough?

That is why I originally said "you need to add a size field" :)

levitte replied:
Yeah, but like I said, this issue is becoming bigger than this PR, so wont fix here.

richsalz replied:
https://github.com/openssl/openssl/issues/9458

levitte commented on crypto/err/err_locl.h:
```diff
@@ -0,0 +1,52 @@
+        err_clear_data(p, i);                                   \
+        (p)->err_flags[i] = 0;                                  \
+        (p)->err_buffer[i] = 0;                                 \
+        (p)->err_file[i] = (NULL);                              \
```
The pain is that I forgot to parenthesize the variables...

levitte commented on crypto/provider_core.c:
```diff
@@ -718,6 +717,17 @@ static const OSSL_PARAM param_types[] = {
 };
 
+/* Check correct function definition */
+static OSSL_core_get_param_types_fn core_get_param_types;
```
If the definition further down uses a different signature, the compiler will complain. Maybe a better comment...

levitte commented on doc/man3/ERR_new.pod:
```diff
@@ -0,0 +1,77 @@
+
+=head1 DESCRIPTION
+
+The functions described here are not generally not used directly, but
```
Will do

levitte commented on doc/man3/ERR_new.pod:
```diff
@@ -0,0 +1,77 @@
+=head1 DESCRIPTION
+
+The functions described here are not generally not used directly, but
+rather through macros that use them, such as L<ERR_raise(3)>, but can
```
Okie

levitte commented on doc/man3/ERR_new.pod:
```diff
@@ -0,0 +1,77 @@
+record.
+
+ERR_vset_error() works like ERR_set_error(), but takes a B<va_list>
+argument instead of the arbitrary number of them.
```
Okie

levitte commented on doc/man3/ERR_put_error.pod:
```diff
@@ -2,31 +2,38 @@
+
 =head1 DESCRIPTION
 
+ERR_raise() raises adds new error to the thread's error queue.  The
```
"adds a"

levitte commented on doc/man3/ERR_put_error.pod:
```diff
@@ -2,31 +2,38 @@
+ERR_raise() raises adds new error to the thread's error queue.  The
+error occured in the library B<lib> for the reason given by the
+B<reason> code.  Additional information can be given with the format
+string B<fmt> and an arbitrary amount of values to be printer, which
```
Thanks

levitte commented on doc/man3/ERR_put_error.pod:
```diff
@@ -2,31 +2,38 @@
+B<reason> code.  Additional information can be given with the format
+string B<fmt> and an arbitrary amount of values to be printer, which
+are processed with L<BIO_snprintf(3)>.  Furthermore, the name of the
+file, the line, and name of the function where that the error occured
```
Thanks

levitte commented on doc/man3/ERR_put_error.pod:
```diff
@@ -2,31 +2,38 @@
+string B<fmt> and an arbitrary amount of values to be printer, which
+are processed with L<BIO_snprintf(3)>.  Furthermore, the name of the
+file, the line, and name of the function where that the error occured
+in is saved with the error record, in so far that the build supports
```
Okie

levitte commented on doc/man7/provider-base.pod:
```diff
@@ -129,25 +131,47 @@ parameters.
 the reason strings table that's returned by
 provider_get_reason_strings().
+The additional data is given as a format string I<fmt> and a set of
+arguments I<args>, which are supposed to be treated in the same manner
```
Okie

levitte commented:
I've a question...  Right now, this PR implements `ERR_raise`, which always requires a format string or NULL, i.e. you have to call it like this at a minimum:

``` C
    ERR_raise(ERR_LIB_PROV, ERR_R_MALLOC_FAILURE, NULL);
```

Is that dangling NULL ok, or will people find that jarring?  Should I then implement two variants, `ERR_raise` that simply takes two arguments, and `ERR_raise_data` that takes a formatting string and arguments?  That would mean this simply case:

``` C
    ERR_raise(ERR_LIB_PROV, ERR_R_MALLOC_FAILURE);
```

and this for the more complex report:

``` C
    ERR_raise_data(ERR_LIB_PROV, ERR_R_MALLOC_FAILURE, "name=%s", name);
```

t8m commented:
I'd vote for the cleanliness approach with two functions. At least in case we expect to have many calls that do not need to supply any additional data - and I suppose it would be so because currently almost all the calls do not supply it.

beldmit commented:
Sorry for brevity, can't browse the PR.

I'd prefer to have not an integer library code but a string one.

levitte commented:
@beldmit, there are associated strings with the library codes, and for those who want to quickly check what library an error came from, having to strcmp is a bit overkill...

levitte commented:
@t8m, that's what I was thinking.  So ok, I'll make that change.

paulidale commented:
We've got a fast string to small integer map, so converting from a string library name to an integer one wouldn't be a major imposition (plus these are errors which should be uncommon).  I agree with @beldmit that it would be nicer.

Doing the same for the reason would be trickier (typos) but it would sidestep the one vs two API question.

I'd love to see: `ERR_raise(const char *library, const char *reason, ...)` with printf formatting escapes in the reason.  A quixotic dream I know.

levitte commented:
> We've got a fast string to small integer map, so converting from a string library name to an integer one wouldn't be a major imposition (plus these are errors which should be uncommon). I agree with @beldmit that it would be nicer.

That will still require that the caller does something more than just compare `ERR_LIB(num) == ERR_LIB_RSA`...  the namemap is intended for *dynamically* generated numeric identities, so they will somehow have to find the number for "RSA" before they can compare.  That puts a requirement on our users that doesn't quite fit with the idea of being source API compatible with pre-3.0 for things that exist in pre-3.0

levitte commented:
In other words, I believe that the library and reason as strings is for some OpenSSL > 3.0.

levitte committed:
ERR: re-use the err_data field when possible

To deallocate the err_data field and then allocating it again might be
a waste of processing, but may also be a source of errors when memory
is scarce.  While we normally tolerate that, the ERR sub-system is an
exception and we need to pay closer attention to how we handle memory.

This adds a new err_data flag, ERR_TXT_IGNORE, which means that even
if there is err_data memory allocated, its contents should be ignored.
Deallocation of the err_data field is much more selective, aand should
only happen when ERR_free_state() is called.

Fixes #9458

Reviewed-by: Paul Dale <paul.dale@oracle.com>
(Merged from https://github.com/openssl/openssl/pull/9459)

levitte commented:
Meanwhile, the CIs are finally happy

paulidale reviewed (dismissed):
If the name isn't changing and the macros aren't refactored into functions....

levitte commented:
> If the name isn't changing and the macros aren't refactored into functions....

Considering other ERR branches just went in, I need to rebase and refactor a little bit anyway, so...

paulidale commented:
I'll wait for the new version.

levitte committed:
fixup! ERR: Add new building blocks for reporting errors

richsalz commented:
If you want two functions, make the "shorter" one a macro:
```
#define ERR_raise(l, r) ERR_raise_data((l), (r), NULL)
```

levitte commented:
> If you want two functions, make the "shorter" one a macro:
> ```
> #define ERR_raise(l, r) ERR_raise_data((l), (r), NULL)
> ```

Isn't that *exactly* what I've done?

https://github.com/openssl/openssl/pull/9452/files#diff-9a5faf9ca0ffd5af12ebe8ea072ff981R249

richsalz commented:
Oh yes, you did that, sorry.  I was reading the comment threads to see what I’d have to do for #9441.

levitte commented:
Merged.

e039ca38c8d77f1e2f182123727c884aaf2d683d Move some macros from include/openssl/opensslconf.h.in, add OPENSSL_FUNC
8a4dc425cc73040c55bc01d89c5541e37dab939a ERR: refactor useful inner macros to err_locl.h.  Add function name field
7c0e20dc6f11aa506abc99ccc90b3a39c48c3052 ERR: Add new building blocks for reporting errors
ed57f7f93508776b898e4c23b65d67f3479edaf1 ERR: Implement the macros ERR_raise() and ERR_raise_data() and use them
add8c8e9647a71cc755dea22490e2075e342624b ERR: Remove ERR_put_func_error() and reimplement ERR_put_error() as a macro
49c6434673ca5e9413062851979cf6ed126c9f1c Refactor provider support for reporting errors
036913b1076da41f257c640a5e6230476c647eff Adapt the FIPS provider to use the new core error functions
c3612970465d0a13f2fc5b47bc28ca18516a699d Avoid using ERR_put_error() directly in OpenSSL code
